common/src/main/java/org/conscrypt/AbstractConscryptSocket.java

```diff
@@ -588,10 +588,10 @@ final PeerInfoProvider peerInfoProvider() {
     abstract void setApplicationProtocolSelector(ApplicationProtocolSelectorAdapter selector);
 
     final void notifyHandshakeCompletedListeners() {
-        if (listeners != null && !listeners.isEmpty()) {
```

Comment:
Make a local reference to the field first and then use that for the tests here and for the argument to the ArrayList copy.